### DISCUSSION TITLE
I have a large file. I want to transfer this file block by block.

### DISCUSSION THREAD

**Discussion by RemonAhammad** [libp2p/rust-libp2p#4213] [Q&A] [answered]
I have a large file. I want to transfer this file block by block (1 MB per transfer) from one device to another. how can I do that using the ibp2p package?

**Comment by thomaseizinger** [comment, marked as answer]
You can certainly do that. You can:

- look into using an existing protocol like https://github.com/ipfs-rust/libp2p-bitswap
- implement your own using `libp2p-request-response` if you want to follow a request-response model
- roll a custom protocol by implementing `NetworkBehaviour` and `ConnectionHandler` if you e.g. want to use more of a streaming approach

For the most part, `rust-libp2p` deals with the abstractions around managing connections for you and running protocols over them concurrently. A very simple example of a file-sharing protocol is in https://github.com/libp2p/rust-libp2p/tree/master/examples/file-sharing but that one is not using blocks.

If you want to implement your own protocol, I would suggest to first define on paper, what your messages look like. Once that is laid out, we can help you in mapping such a protocol to our abstractions.
